crates/moss-cache/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+use dashmap::DashMap;
+use moka::sync::{Cache, CacheBuilder};
+use moka::Expiry;
+use std::any::{type_name, Any};
+use std::fmt::Debug;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use thiserror::Error;
+
+type DynType = Arc<dyn Any + Send + Sync>;
+
+#[derive(Debug, Error)]
+pub enum MossCacheError {
+    #[error("The value of key '{key}' does not have type '{type_name}'")]
+    TypeMismatch { key: String, type_name: String },
+
+    #[error("The key '{key}' does not exist")]
+    NonexistentKey { key: String },
+}
+
+#[derive(Clone, Debug)]
+struct MossCacheItem {
+    pub item: DynType,
+    ttl: Duration,
+}
+
+impl MossCacheItem {
+    pub(crate) fn get<T>(&self, key: &str) -> Result<Arc<T>, MossCacheError>
+    where
+        T: Any + Send + Sync,
+    {
+        match self.item.clone().downcast::<T>() {
+            Ok(result) => Ok(result),
+            Err(_) => Err(MossCacheError::TypeMismatch {
+                key: key.to_string(),
+                type_name: type_name::<T>().to_string(),
+            }),
+        }
+    }
+}
+
+pub struct CustomExpiry {}
+impl Expiry<String, MossCacheItem> for CustomExpiry {
+    fn expire_after_create(
+        &self,
+        key: &String,
+        value: &MossCacheItem,
+        created_at: Instant,
+    ) -> Option<Duration> {
+        Some(value.ttl)
+    }
+
+    fn expire_after_read(
+        &self,
+        key: &String,
+        value: &MossCacheItem,
+        read_at: Instant,
+        duration_until_expiry: Option<Duration>,
+        last_modified_at: Instant,
+    ) -> Option<Duration> {
+        Some(value.ttl)
+    }
+}
+
+impl MossCacheItem {
+    fn new(item: DynType, ttl: Duration) -> MossCacheItem {
+        MossCacheItem { item, ttl }
+    }
+}
+
+type DynamicTypeCache = Cache<String, MossCacheItem>;
+
+pub struct MossCacheBuilder {
+    cache_builder: CacheBuilder<String, MossCacheItem, DynamicTypeCache>,
+}
+
+impl MossCacheBuilder {
+    pub fn new(max_capacity: u64) -> MossCacheBuilder {
+        MossCacheBuilder {
+            cache_builder: CacheBuilder::new(max_capacity).expire_after(CustomExpiry {}),
+        }
+    }
+
+    pub fn build(self) -> MossCache {
+        MossCache {
+            cache: self.cache_builder.build(),
+            retrievers: DashMap::new(),
+        }
+    }
+}
+
+pub struct MossCache {
+    cache: Cache<String, MossCacheItem>,
+    retrievers: DashMap<String, Arc<dyn Fn() -> MossCacheItem>>,
+}
```

Comment:
Why do we need MossCacheBuilder? Can we avoid having an intermediate type in this case?

Reply:
I think so. We can just provide a MossCache::new()